### PR TITLE
Document editable packages

### PR DESCRIPTION
**Pull request by jgsogo**
Docs for https://github.com/conan-io/conan/pull/4181

Light one, as some points will change as soon as we receive some feedback

### CODE REVIEW

File: mastering/editable_packages.rst [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+create **a link from the package in the cache to the local working directory**:
+
+.. code-block:: bash
+
+    $ conan link <path/to/local/dev/libcool> cool/version@user/dev
+
+and we need a file to tell Conan what is the layout of your local project in order to provide the
+proper paths to the build system. There are several ways to achieve this objective:
+
+ * A file next to the *conanfile.py* with the name *.conan_layout*:
```
Comment (danimtb) [inline mastering/editable_packages.rst]:
```suggestion
 * Creating a *.conan_layout* next to next to the *conanfile.py*:
```

File: mastering/editable_packages.rst [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+Setting it all up
+-----------------
+
+The package ``cool/version@user/dev`` is already working, the developer has the sources in a
+local folder, he/she is using whatever method to build and develop locally and can perform
+a :command:`conan create . libCool/version@user/dev` to create the package.
+
+Also, there is a *conanfile.txt* (or a more complex recipe) for the application ``CoolApp`` that
+has ``cool/version@user/dev`` among its requirements. When building this application, the
+resources of ``cool`` are used from the *data* directory inside the Conan local cache.
```
Comment (danimtb) [inline mastering/editable_packages.rst]: No need to specify that the cache is actually un *data*
```suggestion
resources of ``cool`` are used from the Conan local cache.
```

File: mastering/editable_packages.rst (hunk 1 above)
Comment (danimtb) [reply mastering/editable_packages.rst]: Don't you have to have a *cloned* repository with the sources? I think stating that explicitly is important

File: mastering/editable_packages.rst [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+       src/core/include
+       src/cmp_a/include
+
+       [libdirs]
+       build/{settings.build_type}/{options.shared}
+
+       [bindirs]
+       build/{settings.build_type}/{options.shared}
+
+ * A file inside the Conan cache in the path *layouts/default*. As this file can contain information
```
Comment (danimtb) [inline mastering/editable_packages.rst]:
```suggestion
 * Creating a layout file *layouts/default* inside the Conan cache. As this file can contain information
```

File: mastering/editable_packages.rst [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+       [*:libdirs]
+       build/{settings.build_type}/{options.shared}
+
+       [*:bindirs]
+       build/{settings.build_type}/{options.shared}
+
+
+   This file can be very handy inside a company where all the packages have the same layout.
+
+As you can see, you can use inside these files some **placeholders** that will be substituted with
```
Comment (danimtb) [inline mastering/editable_packages.rst]:
```suggestion
As you can see, you can use some **placeholders** inside these files that will be substituted with
```

File: mastering/editable_packages.rst [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+the values of the ``settings`` and the ``options`` of the package.
+
+About the precedence if both files are present, Conan will use the one inside the repository if
```
Comment (danimtb) [inline mastering/editable_packages.rst]:
```suggestion
Regarding the precedence if both files are present, Conan will use the one inside the repository if
```

File: mastering/editable_packages.rst [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+present, then the one in the Conan cache, and if none is available, Conan will fallback to the
+directories defined in the ``package_info`` method of the recipe itself but related to the path
```
Comment (danimtb) [inline mastering/editable_packages.rst]:
```suggestion
directories defined in the ``package_info`` method of the recipe itself but relative to the path
```

File: mastering/editable_packages.rst [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+where the editable package has been linked to.
+
```
Comment (danimtb) [inline mastering/editable_packages.rst]:
```suggestion
```

File: mastering/editable_packages.rst [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+
+Using a package in editable mode
+--------------------------------
+
+Once a package is in editable mode it is used **system wide** by Conan (by every Conan client that
```
Comment (danimtb) [inline mastering/editable_packages.rst]: I think here you can also say that the editable binaries will be used regardless of the profile settings (unless you use hte suitable placeholders)

File: mastering/editable_packages.rst [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+uses the same cache), no changes are required in the consumers. Every :command:`conan install`
+command that requires our editable ``cool/version@user/dev`` package will use the paths to
+the local directory and the changes made to this project will be taken into account by the
+packages using its headers or linking against it.
+
```
Comment (danimtb) [inline mastering/editable_packages.rst]:
```suggestion
```

File: mastering/editable_packages.rst [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+
+Revert the editable mode
+------------------------
+
+In order to revert the editable mode, only the link has to be removed, there is command for that:
```
Comment (danimtb) [inline mastering/editable_packages.rst]:
```suggestion
In order to revert the editable mode just remove the link using:
```

File: mastering/editable_packages.rst [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+
+.. code-block:: bash
+
+    $ conan link --remove cool/version@user/dev
+
+It will remove the link (the local directory won't be affected) and all the packages using this
+requirement will use the one in the cache again.
```
Comment (danimtb) [inline mastering/editable_packages.rst]:
```suggestion
requirement will use the packages in the cache again.
```

File: mastering/editable_packages.rst [hunk 12]
```diff
@@ -0,0 +1,131 @@
...
+
```
Comment (danimtb) [inline mastering/editable_packages.rst]:
```suggestion
```

File: mastering/editable_packages.rst [hunk 13]
```diff
@@ -0,0 +1,131 @@
...
+.. warning::
+
+   Packages that are built using an editable package in its graph upstreams can generate binaries
+   and packages incompatible with the released version of the editable package. Avoid uploading
+   these packages without re-creating them with the in-cache version of all the libraries.
+
+
+.. note::
+
+   This is an experimental feature and it is in its early stage, it can suffer changes on its
```
Comment (danimtb) [inline mastering/editable_packages.rst]: Place this warning at the top of the section like the other ones to warn users from the beginning

File: mastering/editable_packages.rst (hunk 2 above)
Comment (jgsogo) [reply mastering/editable_packages.rst]: I was thinking about the _Conan home_... yes, the _cache_ is actually the _data_ directory.

File: mastering/editable_packages.rst [hunk 14]
```diff
@@ -0,0 +1,131 @@
...
+Also, there is a *conanfile.txt* (or a more complex recipe) for the application ``CoolApp`` that
+has ``cool/version@user/dev`` among its requirements. When building this application, the
+resources of ``cool`` are used from the *data* directory inside the Conan local cache.
+
+
+Put a package in editable mode
+------------------------------
+
+To avoid creating the package ``cool/version@user/dev`` in the cache for every change, we can
+create **a link from the package in the cache to the local working directory**:
```
[the remark this answers is not in the document]
Comment (jgsogo) [reply mastering/editable_packages.rst]: I think it this way: you go from the cache to the local folder... ([+info](https://news.ycombinator.com/item?id=1984456))

File: mastering/editable_packages.rst [hunk 15]
```diff
@@ -0,0 +1,131 @@
...
+
+.. code-block:: bash
+
+    $ conan link <path/to/local/dev/libcool> cool/version@user/dev
+
+and we need a file to tell Conan what is the layout of your local project in order to provide the
```
[the remark this answers is not in the document]
Comment (jgsogo) [reply mastering/editable_packages.rst]: I have modified a couple of things around this line (before the code-block and after) in the next commit, I think that now the present tense fits better.

File: mastering/editable_packages.rst (hunk 1 above)
Comment (jgsogo) [reply mastering/editable_packages.rst]: No, a repository is not needed, just the `conanfile.py` and this file (if you are using it).

File: mastering/editable_packages.rst (hunk 8 above)
Comment (jgsogo) [reply mastering/editable_packages.rst]: Is it not understood from the next sentence: _"Every `conan install`
command that requires our editable ``cool/version@user/dev`` package will use the paths to
the local directory"_

File: mastering/editable_packages.rst (hunk 14 above)
[the remark this answers is not in the document]
Comment (danimtb) [reply mastering/editable_packages.rst]: ok, but I would delete "the package" as this could imply that you actually can have different packages of the same reference in editable mode

File: mastering/editable_packages.rst (hunk 11 above)
Comment (jgsogo) [reply mastering/editable_packages.rst]: rephrased

File: mastering/editable_packages.rst (hunk 8 above)
Comment (danimtb) [reply mastering/editable_packages.rst]: I understood it but I suspect we are a bit biased and it is better to be explicit

Comment (jgsogo) [reply mastering/editable_packages.rst]: I think that bias is thinking about different profile settings 😆 

But, yes, I will add a line there

**Comment by danimtb** [comment]
Fine then! Look at the unresolved suggestions above

File: mastering/editable_packages.rst [hunk 16]
```diff
@@ -0,0 +1,125 @@
...
+to create **a link from the reference in the cache to the local working directory**:
+
+.. code-block:: bash
+
+    $ conan link <path/to/local/dev/libcool> cool/version@user/dev
+
+and all we need is a file to tell Conan what is the layout of our local project in order to provide
+the proper paths to the build system. There are several ways to achieve this objective:
+
+ * A file next to the *conanfile.py* with the name *.conan_layout*:
```
Comment (SSE4) [inline mastering/editable_packages.rst]: previous sentence talks about `conanfile.txt`, but now its `conanfile.py`, which is a bit confusing. is it a different conanfile from the previously mentioned?

File: mastering/editable_packages.rst [hunk 17]
```diff
@@ -0,0 +1,125 @@
...
+
+
+   This file can be very handy inside a company where all the packages have the same layout.
+
+As you can see, you can use some **placeholders** inside these files that will be substituted with
+the values of the ``settings`` and the ``options`` of the package.
+
+Regarding the precedence if both files are present, Conan will use the one inside the repository if
+present, then the one in the Conan cache, and if none is available, Conan will fallback to the
+directories defined in the ``package_info`` method of the recipe itself but relative to the path
```
Comment (SSE4) [inline mastering/editable_packages.rst]: which directories from `package_info`? includedirs? libdirs? both? something else?

File: mastering/editable_packages.rst [hunk 18]
```diff
@@ -0,0 +1,125 @@
...
+
+and all we need is a file to tell Conan what is the layout of our local project in order to provide
+the proper paths to the build system. There are several ways to achieve this objective:
+
+ * A file next to the *conanfile.py* with the name *.conan_layout*:
+
+   .. code-block:: ini
+
+       [includedirs]
+       src/core/include
```
Comment (SSE4) [inline mastering/editable_packages.rst]: for me its not obvious for which directory sub-directories like `src` and `build` are relative to?
to the `conanfile.txt` (`conanfile.py`)? or to the linked directory? or to the current directory?

File: mastering/editable_packages.rst (hunk 16 above)
Comment (jgsogo) [reply mastering/editable_packages.rst]: The only reference to a `conanfile.txt` appears in the previous section asociated with the application `CoolApp`:

> Also, there is a conanfile.txt (or a more complex recipe) for the application CoolApp that has cool/version@user/dev among its requirements. When building this application, the resources of cool are used from the Conan local cache.

In the line you are pointing to I'm talking about the package in editable mode, `cool/version@user/dev`. Nevertheless I'm going to change the line:

> A file next to the conanfile.py with the name .conan_layout:

by

> A file next to the conanfile.py (of the package `cool`) with the name .conan_layout:

It will be less confusing this way?

Comment (SSE4) [reply mastering/editable_packages.rst]: yeah, it's much less confusing now which conanfile are you referring to.

File: developing_packages/editable_packages.rst [hunk 19]
```diff
@@ -0,0 +1,224 @@
...
+Nevertheless, in some case, it is useful to work in several libraries at the same time and see how
+the changes in one of them are propagated to the others. Following the
+:ref:`local workflow<package_dev_flow>` an user can execute the commands :command:`conan source`,
+:command:`conan install`, :command:`conan build` and :command:`conan package`, but in order to
+get the changes ready for a consumer library, it is needed the :command:`conan create` that will
+actually trigger a build to generate the binaries in the cache or to run :command:`conan export-pkg`
+to copy locally built artifacts into the conan cache and make them available to consumers.
+
+What about if you just can tell Conan where to find the headers and the artifacts ready for
+consumption in your local working directory? Not need to package, just tell Conan to use those
```
Comment (jgsogo) [inline developing_packages/editable_packages.rst]:
```suggestion
consumption in your local working directory? No need to package, just tell Conan to use those
```

File: developing_packages/editable_packages.rst [hunk 20]
```diff
@@ -0,0 +1,224 @@
...
+path can be absolute too). If it is found, that will be used. It is possible to add those
+layouts in the source repositories, so they are always easy to find after a clone.
+
+If the specified layout is not found relative to the current directory, it will be looked
+for in the conan cache, in the ``.conan/layouts`` folder. This is very convenient to have
+a single definition of layouts that can be shared with the team and installed with
+``conan config install``.
+
+If no argument is specified, the :command:`conan link` command will try to use a `.conan/layouts/default`
+layout from the local cache.
```
Comment (jgsogo) [inline developing_packages/editable_packages.rst]: I would add a comment about when the `Recipe defined layout` will be used if there is some kind of default to `.conan/layouts/default` file

File: developing_packages/editable_packages.rst [hunk 21]
```diff
@@ -0,0 +1,224 @@
...
+``cool/version@user/dev`` package will use the paths to the local directory and the changes
+made to this project will be taken into account by the packages using its headers or linking
+against it. 
+
+To summarize, consumption of packages in editable mode is transparent to their consumers.
+To try that it is working, the following flow should work:
+
+- get sources of ``cool/version@user/dev``: :command:`git/svn clone... && cd folder`
+- Put package in editable mode: :command:`conan link . cool/version@user/dev --layout=mylayout`
+- Build it usin the local flow: :command:`conan install` + build
```
Comment (jgsogo) [inline developing_packages/editable_packages.rst]:
```suggestion
- Build it using the local flow: :command:`conan install` + build
```

File: developing_packages/editable_packages.rst [hunk 22]
```diff
@@ -0,0 +1,224 @@
...
+- Go to the consumer project: `CoolApp`
+- Build it usin the local flow: :command:`conan install` + build
```
Comment (jgsogo) [inline developing_packages/editable_packages.rst]:
```suggestion
- Build it using the local flow: :command:`conan install` + build
```

File: developing_packages/editable_packages.rst [hunk 23]
```diff
@@ -0,0 +1,224 @@
...
+made to this project will be taken into account by the packages using its headers or linking
+against it. 
+
+To summarize, consumption of packages in editable mode is transparent to their consumers.
+To try that it is working, the following flow should work:
+
+- get sources of ``cool/version@user/dev``: :command:`git/svn clone... && cd folder`
+- Put package in editable mode: :command:`conan link . cool/version@user/dev --layout=mylayout`
+- Build it usin the local flow: :command:`conan install` + build
+- Go to the consumer project: `CoolApp`
```
Comment (jgsogo) [inline developing_packages/editable_packages.rst]:
```suggestion
- Go to the consumer project: ``CoolApp``
```

File: developing_packages/editable_packages.rst [hunk 24]
```diff
@@ -0,0 +1,224 @@
...
+- Build it usin the local flow: :command:`conan install` + build
+- Go back to ``cool/version@user/dev`` source folder, do some changes, and just build. No conan commands necessary
+- Go to the consumer project: `CoolApp` and rebuild. It should get the changes from the ``cool`` library.
```
Comment (jgsogo) [inline developing_packages/editable_packages.rst]:
```suggestion
- Go to the consumer project ``CoolApp`` and rebuild. It should get the changes from the ``cool`` library.
```

File: reference/config_files/editable_layout.rst [hunk 25]
```diff
@@ -0,0 +1,40 @@
...
+Editable layout files
+=====================
+
+This file contain information consumed by :ref:`editable packages <editable_packages>`. It is
+an *.ini* file listing the directories that Conan should use for the packages that are opened
+in editable mode:
+
+.. code-block:: ini
+
+    # Affects to all packages but cool/version@user/dev
```
Comment (jgsogo) [inline reference/config_files/editable_layout.rst]: Something like `# Affects to all packages but cool/version@user/dev (this reference has its own dedicated section)`?

Or maybe it is better to divide this into two code-blocks:
 * First one with sections without references, just showing relative paths, absolute ones and the placeholders things.
 * Second code-block with references and without to show and explain the behavior

File: reference/config_files/editable_layout.rst [hunk 26]
```diff
@@ -0,0 +1,40 @@
...
+    build/{settings.build_type}/{settings.arch}
+
+The specific sections using a package reference will have higher priority than the general ones.
+
+
+This file can live in the conan cache, in the ``.conan/layouts`` folder, or in a user folder, like
+inside the source repo.
+
+If there exists a ``.conan/layouts/default`` layout file in the cache and no layout file is specified
+in the ``conan link <path> <reference>`` command, that file will be used.
```
Comment (jgsogo) [inline reference/config_files/editable_layout.rst]: Again, when the fallback to the `package_info` is being used.

File: developing_packages/editable_packages.rst [hunk 27]
```diff
@@ -0,0 +1,245 @@
...
+
+Using a package in editable mode
+--------------------------------
+
+Once a reference is in editable mode it is used **system wide** (for every set of ``settings`` and
+``options``) by Conan (by every Conan client that uses the same cache), no changes are
+required in the consumers. Every :command:`conan install` command that requires our editable
+``cool/version@user/dev`` package will use the paths to the local directory and the changes
+made to this project will be taken into account by the packages using its headers or linking
+against it.
```
Comment (SSE4) [inline developing_packages/editable_packages.rst]: I think you previously mentioned that some commands will be unavailable for editable packages (e.g. `conan upload` or `conan export`), right?
if so, may be we can clearly document which commands aren't supported for editable packages here?

File: developing_packages/editable_packages.rst [hunk 28]
```diff
@@ -0,0 +1,250 @@
...
+
+Editable packages layouts
+-------------------------
+
+The custom layout of a package while it is in editable mode can be defined in different ways:
+
+Recipe defined layout
+++++++++++++++++++++++
+
+A recipe can defined a custom layout when it is not living in the local cache, in its ``package_info()`` method,
```
Comment (ericLemanissier) [inline developing_packages/editable_packages.rst]: typo:A recipe can define

Comment (memsharded) [reply developing_packages/editable_packages.rst]: Thanks, fixed in develop branch

File: developing_packages/editable_packages.rst [hunk 29]
```diff
@@ -0,0 +1,250 @@
...
+``cool/version@user/dev`` package will use the paths to the local directory and the changes
+made to this project will be taken into account by the packages using its headers or linking
+against it. 
+
+To summarize, consumption of packages in editable mode is transparent to their consumers.
+To try that it is working, the following flow should work:
+
+- get sources of ``cool/version@user/dev``: :command:`git/svn clone... && cd folder`
+- Put package in editable mode: :command:`conan link . cool/version@user/dev --layout=mylayout`
+- Build it using the local flow: :command:`conan install` + build
```
Comment (solvingj) [inline developing_packages/editable_packages.rst]: Is it necessary to build with the `conan build` command?  The local flow seems completely orthogonal.  For our local development, we build with our IDE (visual studio).

File: developing_packages/editable_packages.rst [hunk 30]
```diff
@@ -0,0 +1,250 @@
...
+                    # NOTE: Use the real layout used in your VS projects, this is just an example
+                    self.cpp_info.libdirs = ["%s_%s" % (self.settings.build_type, self.settings.arch)]
+
+That will define the libraries directories to ``path/to/local/dev/libcool/Release_x86_64``, for example.
+That is only an example, the real layout used by VS would be different.
+
+Layout files
++++++++++++++
+
+Layout files have the following syntax:
```
Comment (ytimenkov) [inline developing_packages/editable_packages.rst]: Would be nice to have some transition phrase or introduction of layout files.
For example:
"Instead of changing recipes it's possible to define layout in a separate file. This is especially useful if you have a large number of components with similar structure".

File: developing_packages/editable_packages.rst (hunk 29 above)
Comment (jgsogo) [reply developing_packages/editable_packages.rst]: Check here: https://github.com/conan-io/docs/pull/1046

File: developing_packages/editable_packages.rst (hunk 30 above)
Comment (jgsogo) [reply developing_packages/editable_packages.rst]: Check here: https://github.com/conan-io/docs/pull/1046
